lib/evmone_precompiles/bls.cpp

```diff
@@ -0,0 +1,57 @@
+#include "bls.hpp"
+#include <blst.hpp>
+
+namespace evmone::crypto::bls
+{
+bool g1_add(uint8_t* _rx, uint8_t* _ry, const uint8_t _x0[64], const uint8_t _y0[64],
+    const uint8_t _x1[64], const uint8_t _y1[64])
+{
+    constexpr auto is_field_element = [](const uint8_t p[64]) {
+        return intx::be::unsafe::load<intx::uint512>(p) < crypto::bls::FieldPrime;
+    };
+
+    if (!is_field_element(_x0))
+        return false;
+    if (!is_field_element(_y0))
+        return false;
+    if (!is_field_element(_x1))
+        return false;
+    if (!is_field_element(_y1))
+        return false;
+
+    blst::blst_fp x0;
+    blst::blst_fp y0;
+    blst::blst_fp x1;
+    blst::blst_fp y1;
+    blst::blst_fp_from_bendian(&x0, &_x0[16]);
+    blst::blst_fp_from_bendian(&y0, &_y0[16]);
+    blst::blst_fp_from_bendian(&x1, &_x1[16]);
+    blst::blst_fp_from_bendian(&y1, &_y1[16]);
+
+    const blst::blst_p1_affine p0_affine{x0, y0};
+    if (!blst::blst_p1_affine_on_curve(&p0_affine))
+        return false;
+
+    const blst::blst_p1_affine p1_affine{x1, y1};
+    if (!blst::blst_p1_affine_on_curve(&p1_affine))
+        return false;
+
+    blst::blst_p1 p0;
+    blst::blst_p1 p1;
+    blst::blst_p1_from_affine(&p0, &p0_affine);
+    blst::blst_p1_from_affine(&p1, &p1_affine);
+
+    blst::blst_p1 out;
+    blst::blst_p1_add_or_double(&out, &p0, &p1);
```

Comment:
There is `blst_p1_add_or_double_affine` so you don't have to convert both points to affine.